File: MarkdownEditor/MarkdownEditor/Sources/Editor/EditorTextFormatter.m
#import "EditorTextFormatter.h"

@implementation EditorTextFormatter {
    TextConverter *_converter;
}

- (instancetype)initWithConverter:(TextConverter *)converter {
    self = [super init];
    if (self) {
        _converter = converter;
    }
    return self;
}

- (NSString *)stringByApplyingFormat:(TextConverterFormat)format
                              toText:(NSString *)text
                               range:(NSRange)range
                       selectedRange:(NSRange *)selectedRange {
    if (range.length == 0) {
        if (selectedRange) {
            *selectedRange = range;
        }
        return text;
    }

    NSString *selectedString = [text substringWithRange:range];
    NSString *formattedString = [_converter formattedStringWithString:selectedString format:format];
    NSMutableString *mutableText = [text mutableCopy];
    [mutableText replaceCharactersInRange:range withString:formattedString];

    if (selectedRange) {
        *selectedRange = NSMakeRange(range.location, formattedString.length);
    }
    return mutableText;
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/Editor/EditorTextFormatter.swift
import Foundation

@objc(EditorTextFormatter)
class EditorTextFormatter: NSObject {
    private let converter: TextConverter

    @objc(initWithConverter:)
    init(converter: TextConverter) {
        self.converter = converter
        super.init()
    }

    @objc(stringByApplyingFormat:toText:range:selectedRange:)
    func stringByApplyingFormat(
        _ format: TextConverterFormat,
        toText text: String,
        range: NSRange,
        selectedRange: UnsafeMutablePointer<NSRange>?
    ) -> String {
        guard range.length > 0, let swiftRange = Range(range, in: text) else {
            selectedRange?.pointee = range
            return text
        }

        let selectedString = String(text[swiftRange])
        let formattedString = converter.formattedString(with: selectedString, format: format)
        var mutableText = text
        mutableText.replaceSubrange(swiftRange, with: formattedString)
        selectedRange?.pointee = NSRange(location: range.location, length: (formattedString as NSString).length)
        return mutableText
    }
}
